payjoin/src/receive/mod.rs

```diff
@@ -575,38 +534,44 @@ impl WantsInputs {
 
     /// Add the provided list of inputs to the transaction.
     /// Any excess input amount is added to the change_vout output indicated previously.
-    pub fn contribute_witness_inputs(
+    pub fn contribute_inputs(
         self,
-        inputs: impl IntoIterator<Item = (OutPoint, TxOut)>,
+        inputs: impl IntoIterator<Item = (PsbtInput, TxIn)>,
     ) -> Result<WantsInputs, InputContributionError> {
         let mut payjoin_psbt = self.payjoin_psbt.clone();
         // The payjoin proposal must not introduce mixed input sequence numbers
         let original_sequence = self
-            .payjoin_psbt
+            .original_psbt
             .unsigned_tx
             .input
             .first()
             .map(|input| input.sequence)
             .unwrap_or_default();
+        let uniform_sender_input_type = self.uniform_sender_input_type()?;
 
         // Insert contributions at random indices for privacy
         let mut rng = rand::thread_rng();
         let mut receiver_input_amount = Amount::ZERO;
-        for (outpoint, txo) in inputs.into_iter() {
-            receiver_input_amount += txo.value;
+        for (psbtin, txin) in inputs.into_iter() {
+            let input_pair = InputPair { txin: &txin, psbtin: &psbtin };
+            let input_type =
+                input_pair.address_type().map_err(InternalInputContributionError::AddressType)?;
+
+            if self.params.v == 1 {
+                // v1 payjoin proposals must not introduce mixed input script types
+                self.check_mixed_input_types(input_type, uniform_sender_input_type)?;
+            }
+
+            receiver_input_amount += input_pair
+                .previous_txout()
+                .map_err(InternalInputContributionError::PrevTxOut)?
+                .value;
             let index = rng.gen_range(0..=self.payjoin_psbt.unsigned_tx.input.len());
-            payjoin_psbt.inputs.insert(
-                index,
-                bitcoin::psbt::Input { witness_utxo: Some(txo), ..Default::default() },
-            );
-            payjoin_psbt.unsigned_tx.input.insert(
-                index,
-                bitcoin::TxIn {
-                    previous_output: outpoint,
-                    sequence: original_sequence,
-                    ..Default::default()
-                },
-            );
+            payjoin_psbt.inputs.insert(index, psbtin);
+            payjoin_psbt
+                .unsigned_tx
+                .input
+                .insert(index, TxIn { sequence: original_sequence, ..txin });
```

Comment:
Copying the `original_sequence` rather than using what's supplied in `txin` might be unexpected behavior.

Ideally BIP 77 should allow mixed sequence in the same way mixed script types are allowed, leaving it up to a client decision. I wonder what @nothingmuch thinks. It seems proper to allow mixed input and mixed sequence in fragment paremeters for v2, but also a completely unnecessary complexity for 80%+ use cases, which I imagine should be permissive rather than restrictive.